confidential-data-hub/hub/src/storage/volume_type/blockdevice/mod.rs

```diff
@@ -3,43 +3,98 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 pub mod error;
-pub mod luks;
 
 use super::SecureMount;
-use crate::secret;
+use crate::{
+    secret,
+    storage::drivers::{
+        filesystem::{FsFormatter, FsType},
+        luks2,
+    },
+};
 use async_trait::async_trait;
+use crypto::rand::random_key;
 use error::{BlockDeviceError, Result};
 use kms::{Annotations, ProviderSettings};
-use log::{debug, error};
+use log::debug;
+use nix::mount::{mount, MsFlags};
 use serde::{Deserialize, Serialize};
 use std::collections::HashMap;
 use strum::{Display, EnumString};
+use tokio::{
+    fs::{symlink, File},
+    io::{AsyncBufReadExt, BufReader},
+};
+use zeroize::Zeroizing;
 
 #[derive(EnumString, Serialize, Deserialize, Display, Debug, PartialEq, Eq)]
 pub enum BlockDeviceEncryptType {
     #[strum(serialize = "luks2")]
-    LUKS2,
+    #[serde(rename = "luks2")]
+    Luks2,
 }
 
 #[derive(Serialize, Deserialize, PartialEq, Debug)]
 pub struct BlockDeviceParameters {
     /// The device number, formatted as "MAJ:MIN".
+    /// This is used to identify the block device.
+    ///
+    /// At least one of `device_id` or `device_path` must be set.
+    /// If both are set, `device_id` will be used.
     #[serde(rename = "deviceId")]
-    pub device_id: String,
+    pub device_id: Option<String>,
+
+    /// The path of the source device path. The data of
+    /// the device will be encrypted.
+    ///
+    /// At least one of `device_id` or `device_path` must be set.
+    /// If both are set, `device_id` will be used.
+    #[serde(rename = "devicePath")]
+    pub device_path: Option<String>,
 
     /// The encryption type. Currently, only LUKS is supported.
     #[serde(rename = "encryptType")]
     pub encryption_type: BlockDeviceEncryptType,
 
     /// Encryption key. If not set, generate a random 4096-byte key
+    /// Legal values are:
+    /// - `true`: Use the key to encrypt/decrypt the block device.
+    /// - `false`: Generate a random key to encrypt/decrypt the block device.
     #[serde(rename = "encryptKey")]
     pub encryption_key: Option<String>,
 
+    /// Format the block device with a filesystem type.
+    /// If not set, the block device will be mounted as a raw device.
+    #[serde(rename = "filesystemType")]
+    pub filesystem_type: Option<FsType>,
+
+    /// Indicates whether the block device is ephemeral.
+    /// If set to `true`, there will perform a filesystem format before mounting, if
+    /// any `filesystem_type` is specified.
+    /// If set to `false`, the block device will be mounted as is.
+    pub ephemeral: Option<String>,
```

Comment:
I wonder if we should call this something more literal, like `format_filesystem`. Whether it is ephemeral storage actually depends on a few different things, such as if we format the fs, how we handle the encryption key (see above option), and how the storage is actually provided by the host. So it might be more clear to name this field more specifically. I don't have strong feelings about this though.

Also not sure this needs to be an option. We can probably just use a bool with a default value.

Reply:
Yes, with ephemeral mounts there will be a new random key. Any existing data will not be readable.

I am just saying that we have more than one option that relate to ephemeral storage. This one and the encryption key parameter above. Thus, maybe this should be called something more specific rather than `ephemeral`.